src/legacy/core_plugins/telemetry/server/fetcher.ts

```diff
@@ -0,0 +1,141 @@
+    return internalRepository;
+  };
+
+  private getCurrentConfigs = async () => {
+    const internalRepository = this.getInternalRepository();
+    const telemetrySavedObject = await getTelemetrySavedObject(internalRepository);
+    const config = this.server.config();
+    const currentKibanaVersion = config.get('pkg.version');
+    const defaultTelemetryUsageFetcher = config.get('telemetry.usageFetcher');
+    const telemetryUrl = getXpackConfigWithDeprecated(config, 'telemetry.url') as string;
+
+    return {
+      telemetryOptIn: getTelemetryOptIn({ currentKibanaVersion, telemetrySavedObject }),
+      telemetryUsageFetcher: getTelemetryUsageFetcher({
+        telemetrySavedObject,
+        defaultTelemetryUsageFetcher,
+      }),
+      telemetryUrl,
+    };
+  };
+
+  private updateLastReported = async () => {
+    const internalRepository = this.getInternalRepository();
+    this.lastReported = Date.now();
+    updateTelemetrySavedObject(internalRepository, {
+      lastReported: this.lastReported,
+    });
+  };
+
+  private checkReportStatus = ({ telemetryOptIn, telemetryUsageFetcher }: any) => {
+    if (telemetryOptIn && telemetryUsageFetcher === 'server') {
+      if (!this.lastReported || Date.now() - this.lastReported > REPORT_INTERVAL_MS) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  private fetchTelemetry = async () => {
+    const { getStats, title } = telemetryCollectionManager.getStatsGetter();
+    this.server.log(['debug', 'telemetry', 'fetcher'], `Fetching usage using ${title} getter.`);
+    const config = this.server.config();
+
+    return await getStats({
+      unencrypted: false,
+      server: this.server,
+      start: moment()
+        .subtract(20, 'minutes')
+        .toISOString(),
+      end: moment().toISOString(),
+      isDev: config.get('env.dev'),
+    });
+  };
+
+  private sendTelemetry = async (url: string, cluster: any): Promise<void> => {
+    this.server.log(['debug', 'telemetry', 'fetcher'], `Sending usage stats.`);
+    await fetch(url, {
+      method: 'post',
+      body: cluster,
+    });
+  };
+
+  private sendIfDue = async () => {
+    if (this.isSending) {
+      return;
+    }
+    try {
+      const telemetryConfig = await this.getCurrentConfigs();
+      if (!this.checkReportStatus(telemetryConfig)) {
+        return;
+      }
+
+      // mark that we are working so future requests are ignored until we're done
+      this.isSending = true;
+      const clusters = await this.fetchTelemetry();
+      for (const cluster of clusters) {
+        await this.sendTelemetry(telemetryConfig.telemetryUrl, cluster);
+      }
+
+      await this.updateLastReported();
+    } catch (err) {
+      this.server.log(
+        ['warning', 'telemetry', 'fetcher'],
+        `Error sending telemetry usage data: ${err}`
+      );
+    }
+    this.isSending = false;
+  };
+
+  public start = () => {
+    this.intervalId = setInterval(() => this.sendIfDue(), this.checkDurationMs);
+  };
```

Comment:
It might be worth considering using `setTimeout` (which reschedules itself) instead of `setInterval` as that would allow us to remove the `isSending` flag entirely.

As we adopt more rxjs streams in our plugins, we can then use their interval which you can configure to only emit when the previous handler has resolved...